lib/src/http_client_request.dart

```diff
@@ -0,0 +1,224 @@
+import 'third_party/cronet/generated_bindings.dart';
+
+/// HTTP request for a client connection.
+///
+/// It handles all of the Http Requests made by [HttpClient].
+/// Provides two ways to get data from the request.
+/// [registerCallbacks] or a [HttpClientResponse] which is a
+/// [Stream<List<int>>]. Either of them can be used at a time.
+///
+/// Example Usage:
+/// ```dart
+/// final client = HttpClient();
+/// client.getUrl(Uri.parse('https://example.com/'))
+///   .then((HttpClientRequest request) {
+///   return request.close();
+/// }).then((HttpClientResponse response) {
+///   // Here you got the raw data.
+///   // Use it as you like.
+/// });
+/// ```
+abstract class HttpClientRequest implements io.IOSink {
+  /// Returns [Future] of [HttpClientResponse] which can be listened for server
+  /// response.
+  ///
+  /// Throws [UrlRequestError] if request can't be initiated.
+  @override
+  Future<HttpClientResponse> close();
+
+  /// This is same as [close]. A [HttpClientResponse] future that will complete
+  /// once the request is successfully made.
+  ///
+  /// If any problems occurs before the response is available, this future will
+  /// completes with an [UrlRequestError].
+  @override
+  Future<HttpClientResponse> get done;
+
+  /// Follow the redirects.
+  bool get followRedirects;
+
+  /// Maximum numbers of redirects to follow.
+  /// Have no effect if [followRedirects] is set to false.
+  int get maxRedirects;
+
+  /// The uri of the request.
+  Uri get uri;
+}
+
+/// Implementation of [HttpClientRequest].
+class HttpClientRequestImpl implements HttpClientRequest {
+  final Uri _uri;
+  final String _method;
+  final Pointer<Cronet_Engine> _cronetEngine;
+  final CallbackHandler _callbackHandler;
+  final Pointer<Cronet_UrlRequest> _request;
+
+  /// Holds the function to clean up after the request is done (if nessesary).
+  ///
+  /// Implemented by: http_client.dart.
+  final void Function(HttpClientRequest) _clientCleanup;
+
+  @override
+  Encoding encoding;
+
+  /// Initiates a [HttpClientRequestImpl]. It is meant to be used by a
+  /// [HttpClient].
+  HttpClientRequestImpl(
+      this._uri, this._method, this._cronetEngine, this._clientCleanup,
+      {this.encoding = utf8})
+      : _callbackHandler =
+            CallbackHandler(wrapper.SampleExecutorCreate(), ReceivePort()),
+        _request = cronet.Cronet_UrlRequest_Create() {
+    // Register the native port to C side.
+    wrapper.RegisterCallbackHandler(
+        _callbackHandler.receivePort.sendPort.nativePort, _request.cast());
+  }
+
+  // Starts the request.
+  void _startRequest() {
+    final requestParams = cronet.Cronet_UrlRequestParams_Create();
+    if (requestParams == nullptr) throw Error();
+    cronet.Cronet_UrlRequestParams_http_method_set(
+        requestParams, _method.toNativeUtf8().cast<Int8>());
+    wrapper.InitSampleExecutor(_callbackHandler.executor);
+
+    final cronetCallbacks = cronet.Cronet_UrlRequestCallback_CreateWith(
```

Comment:
Maybe add a TODO with a reference to the relevant ffigen issue with regard to the casts.